File: python_code/ship.py
import math

class Ship:

    ROTATION_ANGLE = 90
    offsets = None
    center_piece_index = None
    size = None
    sunk = False
    placed = False

    def __init__(self) -> None:
        pass

    def rotate_right(self, rotation) -> None:
        for i in range(len(self.offsets)):
            if i == self.center_piece_index:
                pass
            x, y = self.offsets[i]
            new_x = x * round(math.cos(math.radians(self.ROTATION_ANGLE*rotation))) - y * round(math.sin(math.radians(self.ROTATION_ANGLE*rotation)))
            new_y = x * round(math.sin(math.radians(self.ROTATION_ANGLE*rotation))) + y * round(math.cos(math.radians(self.ROTATION_ANGLE*rotation)))
            self.offsets[i] = (new_x, new_y)

    def damage_ship(self):
        self.size -= 1
        if self.size == 0:
            self.sunk = True

    def is_sunk(self):
        return self.sunk


class ThreePieceShip(Ship):
    
    offsets = [(0,-1),(0,0),(0,1)]
    center_piece_index = 1
    size = 3

    def __init__(self) -> None:
        super().__init__()


class SmallL(Ship):

    offsets = [(0,-1), (0,0), (1,0)]
    center_piece_index = 1
    size = 3

    def __init__(self) -> None:
        super().__init__()


class BigL(Ship):

    offsets = [(0,-2), (0,-1), (0,0), (1,0), (2,0)]
    center_piece_index = 2
    size = 5

    def __init__(self) -> None:
        super().__init__()

File: server/game_logic/ship.js
class Ship {

    constructor(){
        this.ROTATION_ANGLE = 90
        this.offsets = null
        this.centerPieceIndex = null
        this.size = null
        this.sunk = false
        this.placed = false
    }

    rotateRight(rotation){
        for (var i = 0; i < this.offsets.length; i++){
            if (i !== this.centerPieceIndex){
                var x = this.offsets[i].x
                var y = this.offsets[i].y
                var new_x = x * Math.round(Math.cos(this.ROTATION_ANGLE*rotation*(Math.PI/180))) - y * Math.round(Math.sin(this.ROTATION_ANGLE*rotation*(Math.PI/180)));
                var new_y = x * Math.round(Math.sin(this.ROTATION_ANGLE*rotation*(Math.PI/180))) + y * Math.round(Math.cos(this.ROTATION_ANGLE*rotation*(Math.PI/180)));
                this.offsets[i].x = new_x, 
                this.offsets[i].y = new_y;
            }
        }
        return this.offsets;
    }

    damageShip(){
        this.size -= 1;
        if (this.size === 0){
            this.sunk = true;
        }
    }

    isSunk(){
        return this.sunk;
    }

}

export class ThreePieceShip extends Ship {

    constructor(){
        super();
        this.offsets = [{"x":0,"y":-1}, {"x":0,"y":0}, {"x":0,"y":1}];
        this.centerPieceIndex = 1;
        this.size = 3;
    }
}

export class SmallLShip extends Ship {

    constructor(){
        super();
        this.offsets = [{"x":0,"y":-1}, {"x":0,"y":0}, {"x":1,"y":0}];
        this.centerPieceIndex = 1;
        this.size = 3;
    }
}

export class BigLShip extends Ship {

    constructor(){
        super();
        this.offsets = [{"x":0,"y":-2}, {"x":0,"y":-1}, {"x":0,"y":0}, {"x":1,"y":0}, {"x":2,"y":0} ];
        this.centerPieceIndex = 2;
        this.size = 5;
    }
}